Sources/NIOTestUtils/NIOThreadPoolTaskExecutor.swift

```diff
@@ -0,0 +1,92 @@
+///
+/// - warning: Do not escape the task executor from the closure for later use.
+///            This function does not return until the task executor is deinitialized.
+///            Any task using the executor will keep the executor alive. In other words, this
+///            function will not return until all tasks running on the executor are finished.
+///
+/// - warning: In case of cancellation, this function will wait for all tasks running on the
+///            executor to finish. Note that cancellation does not propagate to unstructured tasks.
+///
+/// - Parameters:
+///   - numberOfThreads: The number of threads in the pool.
+///   - body: The closure that will accept the task executor.
+///
+/// - Throws: When `body` throws.
+///
+/// - Returns: The value returned by `body`.
+///
+public func withNIOThreadPoolTaskExecutor<T, Failure>(
+    numberOfThreads: Int,
+    body: (any TaskExecutor) async throws(Failure) -> T
+) async throws(Failure) -> T {
+    let nioThreadPool = NIOThreadPool(numberOfThreads: numberOfThreads)
+    nioThreadPool.start()
+
+    // Use AsyncStream as an easy way to wait for the deinit of the task executor.
+    let (deinitStream, deinitContinuation) = AsyncStream.makeStream(of: Void.self)
+    var deinitStreamIterator = deinitStream.makeAsyncIterator()
+
+    let result: Result<T, Failure>
+    do {
+        let taskExecutor = NIOThreadPoolTaskExecutor(nioThreadPool, deinitContinuation)
+        result = .success(try await body(taskExecutor))
+    } catch {
+        result = .failure(error)
+    }
+
+    // Cancellation shield
+    await Task {
+        // Wait for task executor to deinitialize. This makes sure all tasks are finished.
+        // Our own reference to taskExecutor must be dropped by now in order to not keep it alive.
+        await deinitStreamIterator.next()
+        do {
+            try await nioThreadPool.shutdownGracefully()
+        } catch {
+            fatalError("Failed to shutdown NIOThreadPool")
+        }
+    }.value
+
+    return try result.get()
+}
+
+final class NIOThreadPoolTaskExecutor: TaskExecutor {
+    let nioThreadPool: NIOThreadPool
+    let deinitContinuation: AsyncStream<Void>.Continuation
+
+    init(
+        _ nioThreadPool: NIOThreadPool,
+        _ deinitContinuation: AsyncStream<Void>.Continuation
+    ) {
+        self.nioThreadPool = nioThreadPool
+        self.deinitContinuation = deinitContinuation
+    }
+
+    func enqueue(_ job: consuming ExecutorJob) {
+        let unownedJob = UnownedJob(job)
+        self.nioThreadPool.submit { shouldRun in
+            guard case shouldRun = NIOThreadPool.WorkItemState.active else {
+                // Should never happen. We keep the thread pool running until self is deinitialized.
+                fatalError("Thread pool stopped unexpectedly")
+            }
+            unownedJob.runSynchronously(on: self.asUnownedTaskExecutor())
+        }
+    }
+
+    deinit {
+        self.deinitContinuation.finish()
```

Comment:
We shouldn't use `deinit` for this: this kind of RAII-based cleanup has some nasty side effects when it goes wrong. A better design might be to force structured concurrency here, and say that if Tasks are running on this executor that is an immediate crash.

Reply:
Just to be clear. Do you mean that just trying to run an unstructured Task on the executor should result in a crash, or that we should crash if any Task is still running at the point we want to shutdown the thread pool?

Reply:
The latter.

Reply:
Okay, clear. The reason I didn't do that in the first place was that it's non-deterministic; a task may or may not still be running when we want to shutdown, which may result in a crash only in some situations. I guess we'll need to make that very clear in the documentation, then hopefully it will be fine. I'll have a look at it.

Reply:
Agreed, and as this is testing-only code I think it's acceptable.